api/v1alpha1/module_types.go

```diff
@@ -92,42 +93,112 @@ type KernelMapping struct {
 	Regexp string `json:"regexp"`
 }
 
-// ModuleSpec describes how the OOT operator should deploy a Module on those nodes that need it.
-type ModuleSpec struct {
+type ModprobeArgs struct {
+	// Load is an optional list of arguments to be used when loading the kernel module.
+	// +kubebuilder:validation:MinItems=1
+	Load []string `json:"load,omitempty"`
+
+	// Unload is an optional list of arguments to be used when unloading the kernel module.
+	// +kubebuilder:validation:MinItems=1
+	Unload []string `json:"unload,omitempty"`
+}
 
+type ModprobeSpec struct {
+	// ModuleName is the name of the Module to be loaded.
+	ModuleName string `json:"moduleName"`
+
+	// Parameters is an optional list of kernel module parameters to be provided to modprobe.
+	Parameters []string `json:"parameters,omitempty"`
+
+	// DirName is the root directory for modules.
+	// +kubebuilder:default=/opt
+	DirName string `json:"dirName,omitempty"`
+
+	// Args is an optional list of arguments to be passed to modprobe before the name of the kernel module.
 	// +optional
-	Build *Build `json:"build"`
+	Args *ModprobeArgs `json:"args,omitempty"`
 
+	// If RawArgs are specified, they are passed straight to the modprobe binary; all other properties in this
+	// object are ignored.
 	// +optional
-	// AdditionalVolumes is a list of volumes that will be attached to the DriverContainer / DevicePlugin pod,
-	// in addition to the default ones.
-	AdditionalVolumes []v1.Volume `json:"additionalVolumes"`
+	RawArgs *ModprobeArgs `json:"rawArgs,omitempty"`
+}
 
+type DriverContainerContainerSpec struct {
```

Comment:
I don't think there is a clear, logical, distinction between `DriverContainer` and `DriverContainerSpec`? What is the purpose of this separation?

For example I don't see why `KernelMapping` is a better fit in `DriverContainerContainerSpec` rather than in `DriverContainerSpec` or why `ImageRepoSecret` isn't a part of `DriverContainerContainerSpec` but `ImagePullPolicy` is.

WDYT about having a single object `DriverContainerSpec` for all those fields?

Reply:
I agree it might be a bit confusing.
- `DriverContainerSpec` is `.spec.driverContainer`. Contrary to what the name says, it contains all properties related to the driver **pod**. Right now that pod only contains the container that runs `modprobe` cont, but that might change in the future (maybe?)
- `DriverContainerContainerSpec` is `.spec.driverContainer.container`, which contains properties related to the first container in the driver pod (that runs `modprobe`).

The kernel mappings are in the container properties because they govern the `image` field in there. `ImageRepoSecret` is in `.spec.driverContainer` because the corresponding `.spec.imagePullSecrets` field is part of the [`PodSpec`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core).

Maybe we should find a new name for the driver pod, because calling it DriverContainer is indeed confusing.

Reply:
Technically, a secret will be referenced in the `volume` inside the pod's spec, and then mounted into a filesystem path at the container's spec.

I think it will be more clear to have a single object `spec.driverContainer` to hold all those fields without a separation between pod and container's specs don't you think so?